Civi/Share/CiviMRF/CiviMRFCall.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+namespace Civi\Share\CiviMRF;
+
+use CMRF\Core\AbstractCall;
+use CMRF\Core\Call as CallInterface;
+use CMRF\PersistenceLayer\CallFactory;
+
+class CiviMRFCall extends AbstractCall {
+
+  protected string $requestEntity;
+
+  protected string $requestAction;
+
+  protected array $request;
+
+  protected ?array $reply = NULL;
+
+  protected string $status = CallInterface::STATUS_INIT;
+
+  protected array $metadata = [];
+
+  public static function createNew(
+    string $connector_id,
+    CiviMRFCore $core,
+    string $entity,
+    string $action,
+    array $parameters,
+    ?array $options,
+    ?array $callbacks,
+    CallFactory $factory,
+    string $api_version
+  ): self {
+    if (!is_array($callbacks)) {
+      if (NULL === $callbacks) {
+        $callbacks = [];
+      }
+      else {
+        $callbacks = [$callbacks];
+      }
+    }
+
+    return static::create(
+      $connector_id,
+      $core,
+      $api_version,
+      $entity,
+      $action,
+      $parameters,
+        $options ?? [],
+      $callbacks,
+      $factory
+    );
+  }
+
+  protected static function create(
+    string $connector_id,
+    CiviMRFCore $core,
+    string $api_version,
+    string $entity,
+    string $action,
+    array $parameters,
+    ?array $options,
+    ?array $callbacks,
+    CallFactory $factory
+  ): self {
+    $call = new self($core, $connector_id, $factory);
+
+    // compile request
+    if ('3' === $api_version) {
+      $call->request = $call->compileRequest($parameters, $options);
+      $call->request['entity'] = $entity;
+      $call->request['action'] = $action;
+    }
+    elseif ('4' === $api_version) {
+      $call->request = $parameters;
+    }
+    $call->requestEntity = $entity;
+    $call->requestAction = $action;
+    $call->request['version'] = $api_version;
+    $call->status = CallInterface::STATUS_INIT;
+    $call->metadata['callbacks'] = $callbacks;
+    $call->callbacks = $callbacks;
+
+    return $call;
+  }
+
+  public static function createWithRecord($connector_id, $core, $record, $factory) {
+    // Persisting calls not implemented.
+    return NULL;
+  }
+
+  public function getEntity(): string {
+    return $this->requestEntity;
+  }
+
+  public function getAction(): string {
+    return $this->requestAction;
+  }
+
+  public function getParameters(): array {
+    return $this->extractParameters($this->request);
+  }
+
+  public function getRequest(): array {
+    return $this->request;
+  }
+
+  public function getOptions(): array {
+    return $this->extractOptions($this->request);
+  }
+
+  public function getStatus(): string {
+    return $this->status;
+  }
+
+  /**
+   * @param string $status
+   * @param string $error_message
+   * @param ?string $error_code
+   *
+   * @return void
```

Comment:
I'd prefer a prefer a real php type hint.

Reply:
PHP type hints would make this incompatible with the interface in civimrf_cmrf_abstract_call.

Reply:
If the super class has no return type hint a sub class can add any return type hint because the method signature is still compatible with that of the super class. (Restricting the parameters would of course make the method signature incompatible.)